crates/rolldown/src/ast_scanner/side_effect_detector.rs

```diff
@@ -36,22 +34,39 @@ static SIDE_EFFECT_FREE_MEMBER_EXPR_3: Lazy<FxHashSet<(&'static str, &'static st
 pub struct SideEffectDetector<'a> {
   pub scope: &'a AstScope,
   pub source: &'a Arc<str>,
-  pub attached_comment_vecmap: &'a mut Vec<(Span, bool)>,
+  pub trivias: &'a Trivias,
 }
 
 impl<'a> SideEffectDetector<'a> {
-  pub fn new(
-    scope: &'a AstScope,
-    source: &'a Arc<str>,
-    attached_comment_vecmap: &'a mut Vec<(Span, bool)>,
-  ) -> Self {
-    Self { scope, source, attached_comment_vecmap }
+  pub fn new(scope: &'a AstScope, source: &'a Arc<str>, trivias: &'a Trivias) -> Self {
+    Self { scope, source, trivias }
   }
 
   fn is_unresolved_reference(&mut self, ident_ref: &IdentifierReference) -> bool {
     self.scope.is_unresolved(ident_ref.reference_id.get().unwrap())
   }
 
+  /// Comments containing @__PURE__ or #__PURE__ mark a specific function call
+  /// or constructor invocation as side effect free.
+  ///
+  /// Such an annotation is considered valid if it directly
+  /// precedes a function call or constructor invocation
+  /// and is only separated from the callee by white-space or comments.
+  ///
+  /// The only exception are parentheses that wrap a call or invocation.
+  ///
+  /// <https://rollupjs.org/configuration-options/#pure>
+  /// Copied from https://github.com/oxc-project/oxc/blob/147864cfeb112df526bb83d5b8671b465c005066/crates/oxc_linter/src/utils/tree_shaking.rs#L162-L171
+  pub fn has_pure_notation(&self, span: Span) -> bool {
+    let Some((start, comment)) = self.trivias.comments_range(..span.start).next_back() else {
+      return false;
+    };
+    let span = Span::new(*start, comment.end);
```

Comment:
We could cache the return value of `contains(...)` per `comment.end`. This should do the optimization that these code did. But I wonder the if it's worth.

https://github.com/rolldown/rolldown/blob/9e6437d34e24c0c8f913009ebbb1b25fe4d9a3ce/crates/rolldown/src/ast_scanner/impl_visit.rs#L14-L27

---

If this optimization is worthy, we could create a good first issue for it or implement it in another PR.